src/jit.c

```diff
@@ -0,0 +1,1550 @@
+
+static inline void emit_modrm_reg2reg(struct jit_state *state, int r, int m)
+{
+    emit_modrm(state, 0xc0, r, m);
+}
+
+static inline void emit_modrm_and_displacement(struct jit_state *state,
+                                               int r,
+                                               int m,
+                                               int32_t d)
+{
+    if (d == 0 && (m & 7) != RBP) {
+        emit_modrm(state, 0x00, r, m);
+    } else if (d >= -128 && d <= 127) {
+        emit_modrm(state, 0x40, r, m);
+        emit1(state, d);
+    } else {
+        emit_modrm(state, 0x80, r, m);
+        emit4(state, d);
+    }
+}
+
+static inline void emit_rex(struct jit_state *state, int w, int r, int x, int b)
+{
+    assert(!(w & ~1));
+    assert(!(r & ~1));
+    assert(!(x & ~1));
+    assert(!(b & ~1));
+    emit1(state, 0x40 | (w << 3) | (r << 2) | (x << 1) | b);
+}
+
+/* Emit a REX prefix incorporating the top bit of both src and dst. This step is
+ * skipped if no bits are set.
+ */
+static inline void emit_basic_rex(struct jit_state *state,
+                                  int w,
+                                  int src,
+                                  int dst)
+{
+    if (w || (src & 8) || (dst & 8))
+        emit_rex(state, w, !!(src & 8), 0, !!(dst & 8));
+}
+
+static inline void emit_push(struct jit_state *state, int r)
+{
+    emit_basic_rex(state, 0, 0, r);
+    emit1(state, 0x50 | (r & 7));
+}
+
+static inline void emit_pop(struct jit_state *state, int r)
+{
+    emit_basic_rex(state, 0, 0, r);
+    emit1(state, 0x58 | (r & 7));
+}
+
+static inline void emit_jump_target_address(struct jit_state *state,
+                                            int32_t target_pc)
+{
+    struct jump *jump = &state->jumps[state->num_jumps++];
+    jump->offset_loc = state->offset;
+    jump->target_pc = target_pc;
+    emit4(state, 0);
+}
+#elif defined(__aarch64__)
+static inline uint32_t align_to(uint32_t amount, uint64_t boundary)
+{
+    return (amount + (boundary - 1)) & ~(boundary - 1);
+}
+
+static void emit_instruction(struct jit_state *state, uint32_t instr)
+{
+    assert(instr != BAD_OPCODE);
+    emit_bytes(state, &instr, 4);
+}
+
+/* Get the value of the size bit in most instruction encodings (bit 31). */
+static uint32_t sz(bool sixty_four)
+{
+    return (sixty_four ? UINT32_C(1) : UINT32_C(0)) << 31;
+}
+
+/* [ArmARM-A H.a]: C4.1.64: Add/subtract (immediate).  */
+static void emit_addsub_immediate(struct jit_state *state,
+                                  bool sixty_four,
+                                  enum Opcode op,
+                                  int rd,
+                                  int rn,
+                                  uint32_t imm12)
+{
+    const uint32_t imm_op_base = 0x11000000;
+    emit_instruction(state, sz(sixty_four) | (op << 29) | imm_op_base |
+                                (0 << 22) | (imm12 << 10) | (rn << 5) | rd);
+}
+
+/* [ArmARM-A H.a]: C4.1.67: Logical (shifted register).  */
```

Comment:
You should mention `ArmARM-A H.a` as the following reference:
> https://developer.arm.com/documentation/ddi0487/ha (Arm Architecture Reference Manual for A-profile architecture)

Alternatively, name it "ARM-A" for short.